crates/wasmtime/src/limiter.rs

```diff
@@ -0,0 +1,129 @@
+use crate::{Store, StoreInner};
+use std::rc::Weak;
+
+/// Used by hosts to limit resource consumption of instances.
+///
+/// A [`Store`] can be created with a resource limiter so that hosts can take into account
+/// non-WebAssembly resource usage to determine if a linear memory or table should grow.
+pub trait ResourceLimiter {
+    /// Notifies the resource limiter that an instance's linear memory has been requested to grow.
+    ///
+    /// * `current` is the current size of the linear memory in WebAssembly page units.
+    /// * `desired` is the desired size of the linear memory in WebAssembly page units.
+    /// * `maximum` is either the linear memory's maximum or a maximum from an instance allocator,
+    ///   also in WebAssembly page units. A value of `None` indicates that the linear memory is
+    ///   unbounded.
```

Comment:
To make sure I understand this right, if any `maximum` is in play then this field is set to the minimum of all those maximums? The sources of `maximum` are:

* The actual wasm type itself
* The pooling allocator's configured maximum
* A custom implementation of `RuntimeMemory` has a configured maximum

If my understanding is right, could this be clarified a bit? Basically `None` is only returned if literally nothing is limiting the memory growth, but otherwise `Some` doesn't imply that the wasm type itself had a maximum listed, just that something else somewhere is placing a constraint.

Also, presumably, this function isn't called unless `desired` is already known to be less than `maximum`, right? (or is it called even in those cases?)

Reply:
Your understanding is correct and I can expand upon it in the docs for clarity.

The function is called even when desired exceeds the maximum as some implementers will want to know when the maximum was reached even if it was not directly limited by the limiter.

If that feels a little strange, perhaps a different function with the intention that it is only called when the memory couldn't grow due to it exceeding the maximum (from wasm / pooling allocator / custom runtime memory)?

Reply:
Ah ok nah I think for now this is fine. It runs a possible risk of making it difficult to optimize `memory.grow` or `table.grow` in the future but I don't think that's going to be the bottleneck of anything for a long time.